crossbar/worker/proxy.py

```diff
@@ -126,248 +203,280 @@ def onMessage(self, msg):
-                        "cb_proxy_authextra": msg.authextra,
+                        # for WAMP-cryptosign authentication of the proxy frontend
+                        # to the backend router
                         "pubkey": key['hex'],
-                    }
-                )
-                return backend_session
-
-            def _backend_failed(fail):
-                # print("_backend_failed", fail)
-                self.transport.send(
-                    message.Abort(
-                        "wamp.error.runtime_error",
-                        "Failed to connect to backend: {}".format(fail),
-                    )
-                )
 
-            backend_d = self._controller.map_backend(
-                self,
-                self._realm,
-                self._authid,
-                self._authrole,
-            )
-            backend_d.addCallback(_backend_connected)
-            backend_d.addErrback(_backend_failed)
-        else:
-            # 'msg' is a Deny or a Challenge, so communicate that to
-            # the client before we try connecting to the backend
-            self.transport.send(msg)
-
-    def _do_authentication(self, realm, details):
-        if not realm:
-            return types.Deny(
-                ApplicationError.NO_SUCH_REALM,
-                message='no realm requested',
-            )
-        if not self._controller.has_realm(realm):
-            return types.Deny(
-                ApplicationError.NO_SUCH_REALM,
-                message='no realm "{}" exists on this proxy'.format(realm),
-            )
-        if not self._controller.can_map_backend(details.pending_session, realm, details.authid, details.authrole):
-            # XXX maybe want better errors here? maybe can_map should return a Deny?
-            return types.Deny(
-                "wamp.error.runtime_error",
-                message='Proxy cannot map session to a backend (realm="{}" role="{}")'.format(
-                    realm,
-                    details.authrole,
+                        # forward authentication credentials of the connecting client
+                        #
+                        # the following are the effective (realm, authid, authrole) under
+                        # which the client (proxy frontend connection) was successfully
+                        # authenticated (using the authmethod+authprovider)
+                        "proxy_realm": accept.realm,
+                        "proxy_authid": accept.authid,
+                        "proxy_authrole": accept.authrole,
+                        "proxy_authmethod": accept.authmethod,
+                        "proxy_authprovider": accept.authprovider,
+
+                        # this is the authextra returned from the frontend authenticator, which
+                        # would normally be returned to the client
+                        "proxy_authextra": accept.authextra,
+                    }
```

Comment:
ah, nice .. I guess this can also ultimately allow someone doing `dynamic` auth on the actual-backend to get this information direct from the "real" client.

Reply:
yeah, I've further expanded the info. this is the `authextra` an authenticating (frondend) client receives now:

```
{
   'x_cb_node': 'node1',
   'x_cb_worker': 'worker001',
   'x_cb_peer': 'unix:None',
   'x_cb_pid': 19998,
   'x_cb_proxy_node': 'node1',
   'x_cb_proxy_worker': 'worker003',
   'x_cb_proxy_peer': 'tcp4:127.0.0.1:36902',
   'x_cb_proxy_pid': 20005
}
```

so both the frontend proxy and backend proxy contributed to the `authextra` ultimately returned to the client.

Reply:
![Screenshot from 2020-04-05 18-31-14](https://user-images.githubusercontent.com/233340/78504186-ab750380-776b-11ea-8130-43ed36968052.png)